### PR TITLE
Added an output channel for logging

### PR DESCRIPTION
**Pull request by ngondalia**
This will create an output channel for the trace extension 
in the output panel of vscode. This can be used to provide 
information to users.

Signed-off-by: Neel Gondalia ngondalia@blackberry.com

### CODE REVIEW

**Comment by ngondalia** [comment]
@marco-miller @bhufmann please review

**Comment by marco-miller** [comment]
> @marco-miller @bhufmann please review

I will, asap for me next week though, as I now need to call it a week.

File: vscode-trace-extension/src/extension.ts [hunk 1]
```diff
@@ -6,8 +6,12 @@ import { TraceExplorerAvailableViewsProvider } from './trace-explorer/available-
 import { TraceExplorerOpenedTracesViewProvider } from './trace-explorer/opened-traces/trace-explorer-opened-traces-webview-provider';
```
Comment (marco-miller) [inline vscode-trace-extension/src/extension.ts:6]: A note, say for the future, the `Signed-off-by` footer (commit message, PR description) requires angle brackets around the email. E.g., `<ngondalia@blackberry.com>`.

And, speaking of commit messages, we usually format them [based on these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html). So this could be for future contributions.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+ * Copyright (c) 2023 BlackBerry Limited and others.
+ *
+ * Licensed under the MIT license. See LICENSE file in the project root for details.
+ ***************************************************************************************/
+
+import * as vscode from 'vscode';
+
+export class TraceExtensionLogger {
+    private logChannel: vscode.OutputChannel;
+    private channelDisposed = false;
```
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:11]: I think the `constructor` (below) should explicitly set `channelDisposed` to `false`, while this default being `true` until then.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+     */
+    isCurrentChannelDisposed(): boolean {
+        return this.channelDisposed;
+    }
+
+    /**
+     * Show the log channel from the UI
+     */
+    showLog(): void {
+        if (!this.channelDisposed) {
```
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:30]: Would reuse `isCurrentChannelDisposed()` here instead.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            this.logChannel.show();
+        }
+    }
+
+    /**
+     * Hide the log channel from the UI
+     */
+    hideLog(): void {
+        if (!this.channelDisposed) {
```
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:39]: Same as above.

File: vscode-trace-extension/src/utils/trace-extension-logger.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            this.logChannel.hide();
+        }
+    }
+
+    /**
+     * Clears the entries from the log channel
+     */
+    clearLog(): void {
+        if (!this.channelDisposed) {
```
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:48]: (Same.)

File: vscode-trace-extension/src/utils/trace-extension-logger.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+            this.logChannel.clear();
+        }
+    }
+
+    /**
+     * Disposes the log channel and all of its resources
+     */
+    disposeChannel(): void {
+        this.logChannel.dispose();
```
Comment (marco-miller) [inline vscode-trace-extension/src/utils/trace-extension-logger.ts:57]: If one calls this method more than once in a row, then some `dispose()` implementation might eventually break. Should we check for `isCurrentChannelDisposed()` before calling this here?

File: vscode-trace-extension/src/extension.ts [hunk 7]
```diff
@@ -6,8 +6,12 @@ import { TraceExplorerAvailableViewsProvider } from './trace-explorer/available-
...
 import { fileHandler, openOverviewHandler, resetZoomHandler } from './trace-explorer/trace-tree';
 import { updateTspClient } from './utils/tspClient';
+import { TraceExtensionLogger } from './utils/trace-extension-logger';
+
+export let traceLogger: TraceExtensionLogger;
 
 export function activate(context: vscode.ExtensionContext): void {
+    traceLogger = new TraceExtensionLogger('Trace Extension');
```
Comment (marco-miller) [inline vscode-trace-extension/src/extension.ts:14]: I'm unable to locate this option in the `Developer: Set Log Level` menu of the executing extension (from the command palette), but maybe I'm missing something simply enough. Can you summarize the steps leading to enabling these logs in the Output panel, maybe in the root `README.md`? -With a typical usage example if possible.

**Review by marco-miller — CHANGES_REQUESTED** [review]
Some comments/questions; thanks again for this.

File: README.md [hunk 8]
```diff
@@ -106,6 +106,22 @@ The react-app is another matter. The panel is a webview that is running in its o
 
 Each panel is its own small web application, so to debug, while in the context of the webview, press `ctrl-shift-p` and enter the command `Developer: Open Webview Developer Tools`. This will open the developer tools. The code is in the `Sources` tab of the developer tools window that opens.
 
+### Logging in the extension
+
+The extension uses an output channel for logging. To view the logs, navigate to the output panel. The output panel can be accessed by navigating to view -> open view -> type 'output'. To open the extension output channel, navigate the drop down option and look for `Trace Extension`. An alternate way of opening the trace extension output channel is through command palette. Open command palette by pressing press `ctrl-shift-p`, and then run `Output: Show Output Channels...\`. This will prompt a list of available outputs. Select `Trace Extension` from the list of available outputs.
```
Comment (marco-miller) [inline README.md:111]: `pressing press` is redundant.

File: README.md [hunk 9]
```diff
@@ -106,6 +106,22 @@ The react-app is another matter. The panel is a webview that is running in its o
...
+
+For logging to the `Trace Extension` output channel, use the `traceLogger` object instantiated in extension.ts. The following are examples of using the log channel:
```
Comment (marco-miller) [inline README.md:113]: Would also put `extension.ts` within backticks.

File: README.md [hunk 10]
```diff
@@ -106,6 +106,22 @@ The react-app is another matter. The panel is a webview that is running in its o
...
+
+traceLogger.addLogMessage('Hello from trace extension without tag');
```
Comment (marco-miller) [inline README.md:115]: Format as code, too?

File: README.md [hunk 11]
```diff
@@ -106,6 +106,22 @@ The react-app is another matter. The panel is a webview that is running in its o
...
+
+This will add the following log entry in the output channel:
+[2023-04-25 11:07:22.500] Hello from trace extension without tag
```
Comment (marco-miller) [inline README.md:118]: Would format this other line as code, as well.

File: README.md [hunk 12]
```diff
@@ -106,6 +106,22 @@ The react-app is another matter. The panel is a webview that is running in its o
...
+
+traceLogger.addLogMessage('Hello from trace extension with tag', 'tag');
```
Comment (marco-miller) [inline README.md:120]: Same as above.

File: README.md [hunk 13]
```diff
@@ -106,6 +106,22 @@ The react-app is another matter. The panel is a webview that is running in its o
...
+
+This will add the following log entry in the output channel:
+[2023-04-25 11:08:40.500] [tag] Hello from trace extension with tag
```
Comment (marco-miller) [inline README.md:123]: (Same.)

**Review by marco-miller — CHANGES_REQUESTED** [review]
Only some nits left for me; thanks.

File: README.md [hunk 14]
```diff
@@ -106,6 +106,30 @@ The react-app is another matter. The panel is a webview that is running in its o
 
 Each panel is its own small web application, so to debug, while in the context of the webview, press `ctrl-shift-p` and enter the command `Developer: Open Webview Developer Tools`. This will open the developer tools. The code is in the `Sources` tab of the developer tools window that opens.
 
+### Logging in the extension
+
+The extension uses an output channel for logging. To view the logs, navigate to the output panel. The output panel can be accessed by navigating to view -> open view -> type 'output'. To open the extension output channel, navigate the drop down option and look for `Trace Extension`. An alternate way of opening the trace extension output channel is through command palette. Open command palette by pressing `ctrl-shift-p`, and then run `Output: Show Output Channels...\`. This will prompt a list of available outputs. Select `Trace Extension` from the list of available outputs.
+
+For logging to the `Trace Extension` output channel, use the `traceLogger` object instantiated in `extension.ts`. The following are examples of using the log channel:
+
+```
```
Comment (marco-miller) [inline README.md:115]: \`\`\`javascript, here.

Fenced markdown requires a code language to be specified, as the markdownlint VS Code extension warning otherwise indicates. See [this list](https://www.rubycoloredglasses.com/2013/04/languages-supported-by-github-flavored-markdown/) for more examples.

File: README.md [hunk 15]
```diff
@@ -106,6 +106,30 @@ The react-app is another matter. The panel is a webview that is running in its o
...
+traceLogger.addLogMessage('Hello from trace extension without tag');
+```
+
+This will add the following log entry in the output channel:
+```
```
Comment (marco-miller) [inline README.md:120]: \`\`\`text, here, per above explanation.

File: README.md [hunk 16]
```diff
@@ -106,6 +106,30 @@ The react-app is another matter. The panel is a webview that is running in its o
...
+[2023-04-25 11:07:22.500] Hello from trace extension without tag
+```
+
+```
```
Comment (marco-miller) [inline README.md:124]: \`\`\`javascript

**Review by marco-miller — CHANGES_REQUESTED** [review]
Thanks! -Only missing these fenced language suffixes by now.

File: README.md (hunk 15 above)
Comment (marco-miller) [reply README.md:120]: Addendum, this specific ```text case isn't part of that list linked from above, but can be found elsewhere [as in here](https://github.com/jincheng9/markdown_supported_languages).

**Review by marco-miller — APPROVED** [review]
LGTM now, @bhufmann -back to you for the 2nd review again; thanks.

**Review by bhufmann — APPROVED** [review]
Looks good to me! Thanks for the contribution.
